rundeckapp/grails-app/controllers/rundeck/interceptors/AuditInterceptor.groovy

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2017 Rundeck Inc, <http://rundeck.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package rundeck.interceptors
+
+import com.dtolabs.rundeck.core.audit.ActionTypes
+import com.dtolabs.rundeck.core.audit.ResourceTypes
+import com.dtolabs.rundeck.core.common.FrameworkResource
+import org.springframework.beans.factory.annotation.Autowired
+import rundeck.services.ConfigurationService
+
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * This filter tracks user access to trigger auditing events.
+ * For example, track the user access to different projects.
+ *
+ * This filter assumes {@link ProjectSelectInterceptor} has already processed the request.
+ *
+ * @author Alberto Hormazabal
+ */
+class AuditInterceptor {
+
+    int order = HIGHEST_PRECEDENCE + 101
+
+    /** Default minimum period to wait between project access notifications, in seconds */
+    private static final int DEFAULT_PROJECT_MIN_NOTIFICATION_PERIOD = 60 * 30 // default 30 min.
+
+    /** Default minimum time to keep project access track data on cache, in seconds */
+    private static final int DEFAULT_CACHE_RETENTION_PERIOD = 60 * 30 // default 30 min.
+
+    def frameworkService
+    def auditEventsService
+    final ConfigurationService configurationService
+
+    private Long minProjectNotifyPeriodMillis
+    private Long cacheRetentionPeriodMillis
+
+    private final ConcurrentHashMap<String, ProjectTrackingEntry> projectAccessTrackingMap
```

Comment:
is a Guava cache a better fit here? rather than manually doing cache expiration?

Reply:
I did try first with guava but coulnd't make it work right as i expected. I finally implemented like this to have more control about the atomicity of the map entries updating without creating too much contention between concurrent requests.